freedomofpress/securedrop-client
# scale conversation widgets as window resizes

# Description

Fixes https://github.com/freedomofpress/securedrop-client/issues/1061
Resolves https://github.com/freedomofpress/securedrop-client/issues/1204
Known issues: https://github.com/freedomofpress/securedrop-client/issues/1205, https://github.com/freedomofpress/securedrop-client/issues/1207, https://github.com/freedomofpress/securedrop-client/issues/815, 

Known issue that this PR would introduce: https://github.com/freedomofpress/securedrop-client/issues/1213

# Test Plan

## Basic Test
1. Run the client against a server with some sources
2. Select a source and resize the client window, try double clicking and resizing to the smallest width and height
    - [ ] Confirm there is never horizontal scrollbar in the conversation view
    - [ ] Confirm that content is wrapped within message and reply speech bubbles as you resize the client window (note that https://github.com/freedomofpress/securedrop-client/issues/815 is a known issue)
## Test supported minimum screen resolutions
1. Change your screen resolution to 1280x1024
2. Run through basic testing procedure above
3. Change your screen resolution to 1368x768
4. Run through basic testing procedure above
## Test pending and failed replies
1. Close the client
2. Run this code so that you can create a pending and failed reply in the next steps: `wget https://gist.githubusercontent.com/creviera/7f19a7d10334359f40dbdbb2354cd13a/raw/a2ef94913a155aa4019b753cf916f844c9cffa3a/pending-reply && git apply pending-reply`
3. Log into the client and send a reply to a source, see it pending along with a server connection error message
4. With that source still selected, resize the client a bunch, including to its smallest size, maximize, double-click multiple times, etc.
    - [ ] Confirm there is never horizontal scrollbar in the conversation view
    - [ ] Confirm that content is wrapped within message and reply speech bubbles as you resize the client window (note that https://github.com/freedomofpress/securedrop-client/issues/815 is a known issue)
5. Log out and back in to see the pending reply show up with a failure message
6. Repeat step 4, this time with the failed reply in the conversation view
    - [ ] Confirm there is never horizontal scrollbar in the conversation view
    - [ ] Confirm that content is wrapped within message and reply speech bubbles as you resize the client window (note that https://github.com/freedomofpress/securedrop-client/issues/815 is a known issue)
## Test new messages and replies from the JI
1. Send a message as a source from the JI while the client has the conversation view for that source open
    - [ ] Confirm that the message is displayed in a speech bubble of the correct size no matter the width of the client window (should match the size of other message speech bubbles in the conversation view)
2. Send a reply from the JI to a source that is currently selected in the client you have running
    - [ ] Confirm that the reply is displayed in a speech bubble of the correct size no matter the width of the client window (should match the size of other reply speech bubbles in the conversation view)
## Test file attachments
1. Select a source with a file attachment that has not yet been downloaded
    - [ ] Confirm the download button, "ENCRYPTED FILE ON SERVER" text, and file size all fit within the file widget when the client is scaled down to its minimum width
    - [ ] Confirm the horizontal line between the text and file size does not exist when the client is scaled down to its minimum width
    - [ ] Confirm the horizontal line between the text and file size stretches horizontally as the client window is resized
2. Download the file
    - [ ] Confirm the export/ print buttons, filename, and file size all fit within the file widget when the client is scaled down to its minimum width (note https://github.com/freedomofpress/securedrop-client/issues/1205 is a known issue)
    - [ ] Confirm the horizontal line between the filename and file size does not exist when the client is scaled down to its minimum width (the filename has to be long enough to not see the horizontal line)
    - [ ] Confirm the horizontal line between the text and file size stretches horizontally as the client window is resized

# Checklist

If these changes modify code paths involving cryptography, the opening of files in VMs or network (via the RPC service) traffic, Qubes testing in the staging environment is required. For fine tuning of the graphical user interface, testing in any environment in Qubes is required. Please check as applicable:


If these changes add or remove files other than client code, the AppArmor profile may need to be updated. Please check as applicable:


If these changes modify the database schema, you should include a database migration. Please check as applicable:
